mbtiles/__init__.py

```diff
@@ -47,19 +50,35 @@ def process_tile(tile):
     kwds['transform'] = from_bounds(ulx, lry, lrx, uly, 256, 256)
     src_nodata = kwds.pop('src_nodata', None)
     dst_nodata = kwds.pop('dst_nodata', None)
-
-    with rasterio.open('/vsimem/tileimg', 'w', **kwds) as tmp:
-        reproject(rasterio.band(src, src.indexes),
-                  rasterio.band(tmp, tmp.indexes),
-                  src_nodata=src_nodata,
-                  dst_nodata=dst_nodata,
-                  num_threads=1,
-                  resampling=resampling)
-
-    data = bytearray(virtual_file_to_buffer('/vsimem/tileimg'))
-
-    # Workaround for https://bugs.python.org/issue23349.
-    if sys.version_info[0] == 2 and sys.version_info[2] < 10:
-        data[:] = data[-1:] + data[:-1]
-
-    return tile, data
+    kwds['driver'] = 'GTiff'
+
+    # Reproject into a GeoTIFF tile.
+    with MemoryFile() as memdst:
+        with memdst.open(**kwds) as tiledst:
+            reproject(
+                rasterio.band(src, src.indexes),
+                rasterio.band(tiledst, tiledst.indexes),
+                src_nodata=src_nodata, dst_nodata=dst_nodata, num_threads=2,
+                resampling=resampling)
+
+        # If the tiledst dataset has any valid data, we read and return it.
+        # Otherwise, we return None.
+        #
+        # To save time, we check the only mask of the first band for
+        # valid data.
+        with memdst.open() as tiledst:
```

Comment:
@jdmcbr here's what I was thinking for valid data checking. We've warped the src dataset to a tile-sized GeoTIFF (`tiledst`) that is stored in `memdst`. We can peek at the mask of the first band of `tiledst` to see if there are any valid data pixels. If there are not, we skip the reading of bytes and the worker returns None.

Does that make sense.